### PR TITLE
Beta features, integration tests and  bugfix

### PR DESCRIPTION
**Pull request by mkucherenko**
Test not passed yet:
-testCreateAccessTokenIdempotent
-testAuthFlowTest
-testRecoveryWithSecondaryAgent
-testGetProfilePicture
-testGetPictureProfile

### CODE REVIEW

**Comment by yuhanniu** [comment]
Can we exclude API calls that we know won't work correctly from the failing tests? and also exclude the tests that don't pass.

File: tests/Tokenio/TokenBaseTest.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+use Io\Token\Proto\Common\Address\Address;
+use PHPUnit\Framework\TestCase;
+use Tokenio\Config\TokenCluster;
+use Tokenio\Config\TokenEnvironment;
+use Tokenio\Config\TokenIoBuilder;
+use Tokenio\Util\Strings;
+use Tokenio\Security\UnsecuredFileSystemKeyStore;
+use Io\Token\Proto\Common\Alias\Alias;
+
+abstract class TokenBaseTest extends TestCase
```
Comment (yuhanniu) [inline tests/Tokenio/TokenBaseTest.php]: Is introducing the base test class necessary? We want to keep the implementation approaches similar if possible.

File: lib/Util/Util.php [hunk 2]
```diff
@@ -222,15 +257,40 @@ private static function sha256hash($payload)
...
         if ($json == null) {
             return $data;
         }
-
         self::jsonSort($json);
+        self::fixEmptyObjectReplacements($jsonOriginal, $json, '');
         return json_encode($json);
     }
 
+    private static function fixEmptyObjectReplacements($originalJson, &$object, $path)
```
Comment (yuhanniu) [inline lib/Util/Util.php]: Could you help me understand what you are trying to do with this block? This looks a bit hacky.

Comment (mkucherenko) [reply lib/Util/Util.php]: Yes, that's a little bit hacky, but without it normalizing JSON was broken. When deserializing for sort json by key names and serializing it back we were having empty arrays instead of empty object. There is a way to force serializing those nodes as empty object, but that serializes empty arrays as objects as well. So what it does - simply search for all empty objects in original JSON and makes according changes to result JSON.
image

File: tests/Tokenio/TokenBaseTest.php (hunk 1 above)
Comment (mkucherenko) [reply tests/Tokenio/TokenBaseTest.php]: That class provides base methods like initSDK(), generateAlias(), generateAddress(), and method to convert RepeatedField to array to perform checks. Those could be moved to some kind of TestUtils.php but to me it looks OK to left it in base class. I can change that if needed.

File: lib/Util/Util.php (hunk 2 above)
Comment (yuhanniu) [reply lib/Util/Util.php]: You can use the `JSON_FORCE_OBJECT` option to decode and stop calling ksort when you get an empty object(instead of an array). There has to be a way to implement this correctly without a hack.

File: tests/Tokenio/TokenBaseTest.php (hunk 1 above)
Comment (yuhanniu) [reply tests/Tokenio/TokenBaseTest.php]: I understand, and I don't think this approach itself has a problem. I was trying to say we try to have implementation consistency across multiple SDKs, so we'd rather not use this "base test class" approach since it is not used else where.

You don't need to do it in this PR, but we do need to change it, to use the TestUtil approach like you mentioned. I'll create a ticket for this and we can do this later.

File: lib/Util/Util.php (hunk 2 above)
Comment (mkucherenko) [reply lib/Util/Util.php]: `JSON_FORCE_OBJECT` is what I meant in my previous comment. It forces empty arrays to be converted to empty objects as well, which is also an issue.

Comment (yuhanniu) [reply lib/Util/Util.php]: you are right but there has to be a way to implement this correctly without a hack.

File: lib/Util/Util.php [hunk 3]
```diff
@@ -50,6 +53,17 @@ public static function hashAlias($alias)
...
     }
 
+    /**
+     * Hash Alias as Base58 hashed json from Alias message
+     *
+     * @param Message $message the alias to hash
+     * @return string
+     * @throws \Tokenio\Exception\CryptographicException
+     */
+    public static function hasProto($message)
```
[the remark this answers is not in the document]
Comment (mkucherenko) [reply lib/Util/Util.php]: Fixed

File: lib/Util/Util.php [hunk 4]
```diff
@@ -86,6 +100,25 @@ public static function createAddAliasOperation($alias)
         return $operation;
     }
 
+    /**
+     * Creates MemberOperation from Keys
+     * @param Key[] $keys
+     * @return MemberOperation[]
+     */
+    public static function createMemberOperationsFromKeys($keys)
```
[the remark this answers is not in the document]
Comment (mkucherenko) [reply lib/Util/Util.php]: Fixed

File: tests/Tokenio/TokenBaseTest.php (hunk 1 above)
Comment (mkucherenko) [reply tests/Tokenio/TokenBaseTest.php]: Changed to TestUtil.php

File: lib/Util/Util.php (hunk 2 above)
Comment (mkucherenko) [reply lib/Util/Util.php]: Me and Vlad tried few libraries which do nothing more but wrapper around `json_decode` which doesn't help. So only solution for sorting and normalizing json correctly we came up with is this one.

**Comment by mkucherenko** [comment]
> Can we exclude API calls that we know won't work correctly from the failing tests? and also exclude the tests that don't pass.

I removed 5 tests which are not passing

File: lib/Member.php [hunk 5]
```diff
@@ -292,21 +324,6 @@ public function addAliases($aliasList)
...
 
-    /**
-     * Creates a Representable that acts as another member using the access token
-     * that was granted by that member.
-     *
-     * @param string $tokenId the token id
-     * @param boolean $customerInitiated whether the call is initiated by the customer
-     * @return Member
-     */
-    public function forAccessToken($tokenId, $customerInitiated = false)
```
Comment (yuhanniu) [inline lib/Member.php]: why was this removed?

File: lib/TokenIO.php [hunk 6]
```diff
@@ -258,37 +264,6 @@ public function generateTokenRequestUrl($requestId, $state = '', $csrfToken = ''
...
-     * Parse the token request callback URL to extract the state and the token ID. Verify that the
-     * state contains the CSRF token hash and that the signature on the state and CSRF token is
-     * valid.
-     *
-     * @param string $callbackUrl the token request callback url
-     * @param string $csrfToken the csrf token
-     * @return TokenRequestCallback object containing the token id and the original state
-     * @throws Exception
-     */
-    public function parseTokenRequestCallbackUrl($callbackUrl, $csrfToken = null)
```
Comment (yuhanniu) [inline lib/TokenIO.php]: What's wrong with this call? This call is actually important.
